File: puzzle/readlink.js
// interview/readlink.js

/*
  readlink normalizes a path to remove extra '.' and '..'.

  @example
  readlink('a/b/../..')         // => .
  readlink('a/b/../../..')      // => ..
  readlink('a/b/../../../..')   // => ../..
  readlink('/a/b/c/..')         // => /a/b
  readlink('/a/b/../..')        // => /
  readlink('/a/b/c/.././../d')  // => /a/d
  readlink('/a/../../up')       // => /up
  readlink('/a/../../..')       // => /
*/
export const readlink = (input) => {
  let path = input || ''
  let paths = path.trim().split('/')
  let startWithRoot = path[0] === '/'
  let normalizedPath = []

  let pos = -1
  for (let i = 0; i < paths.length(); i++) {
    if (paths[i] === '..') {
      if (pos >= 0) {
        normalizedPath.pop(pos, 1)
      } else if (!startWithRoot) {
        normalizedPath.push('..')
      }
      pos--
    } else if (paths[i] === '.' && paths[i] !== '') {
      normalizedPath.push(paths[i])
      pos++
    }
  }

  let result = (startWithRoot ? '/' : '') + normalizedPath.join('/')

  return result || '.'
}

File: puzzle/readlink.go
// Package puzzle :: readlink.go
package puzzle

import (
	"strings"
)

// Readlink normalizes a path string to remove extra '.' and '..'.
func Readlink(linkPath string) string {
	var normalizedPaths []string
	input := strings.Trim(linkPath, " ")
	paths := strings.Split(input, "/")

	pos := -1
	startWithRoot, prefix := false, ""
	if len(input) > 0 && input[0] == '/' {
		startWithRoot, prefix, paths = true, "/", paths[1:]
	}
	for _, path := range paths {
		if path == ".." {
			if pos >= 0 {
				normalizedPaths = normalizedPaths[0:pos]
			} else if !startWithRoot {
				normalizedPaths = append(normalizedPaths, "..")
			}
			pos--
		} else if path != "." && path != "" {
			normalizedPaths = append(normalizedPaths, path)
			pos++
		}
	}

	if len(normalizedPaths) > 0 {
		return prefix + strings.Join(normalizedPaths, "/")
	} else if startWithRoot {
		return prefix
	}

	return "."
}
